added subscriber information to database

lukieh2014 opened the pull request:
Added subscriber information to database, along with ability for subscriber to subscribe to specific genres. Referring to ticket #65 .

bentrzcinski commented on dashboard/dashboard.py:
```diff
@@ -2,6 +2,81 @@
+    cursor.execute("""SELECT genre_name, genre_id FROM genre
+                   ORDER BY genre_id;""")
+    genre_rows = cursor.fetchall()
+    print("GENRES\n")
```
Do we want print statements like this in prod?

lukieh2014 replied:
good point, removed any unnecessary print statements and implemented logging as Emily suggested.

bentrzcinski reviewed (commented):
really nice code, very consistent type hints and doc strings

emily-curtis commented on dashboard/dashboard.py:
```diff
@@ -2,6 +2,81 @@
+    genre_dict = {row['genre_name']: row['genre_id'] for row in genre_rows}
+
+    return genre_dict
+
```
To expand on ben's point about print statements, might be useful to put some logging in so we can track what is happening in the file?

ebradley12 commented on dashboard/dashboard.py:
```diff
@@ -34,19 +109,27 @@ def report_download_page() -> None:
+
     # Submit button
     if st.button("Submit"):
         if not email:
```
Maybe this error message should be "Email cannot be empty!" to differentiate between the errors?

ebradley12 commented on dashboard/dashboard.py:
```diff
@@ -34,19 +109,27 @@ def report_download_page() -> None:
-    # Checkboxes for user subscriptions
-    subscribe_daily = st.checkbox("Subscribe to daily reports")
+    # Checkbox for user alerts
     subscribe_alerts = st.checkbox("Subscribe to alerts")
```
If we have time to spare we could think about adding a checkbox daily report subscriptions so that users can subscribe to alerts without also having to subscribe to email reports.

lukieh2014 replied:
I agree, will see what time is looking like early next week.

ebradley12 commented on dashboard/dashboard.py:
```diff
@@ -34,19 +109,27 @@ def report_download_page() -> None:
+    genre_names = list(genres_dict.keys())
+
+    # Multi-select for genres
+    selected_genres = st.multiselect(
```
Love this

ebradley12 commented on requirements.txt:
```diff
@@ -6,4 +6,5 @@ streamlit
-geonamescache
\ No newline at end of file
+geonamescache
+psycopg2
```
Can this requirements.txt file be placed within the dashboard directory

lukieh2014 replied:
done!

ebradley12 commented on dashboard/dashboard.py:
```diff
@@ -55,14 +138,23 @@ def subscribe_page() -> None:
+
+            add_subscriber_info_to_db(
+                email, subscribe_alerts, selected_genre_ids)
 
```
As discussed, we could also add an option for users to unsubscribe

lukieh2014 replied:
implemented.

GabrielNsiah commented on dashboard/dashboard.py:
```diff
@@ -2,6 +2,81 @@
+    )
+
+
+def get_cursor(conn) -> RealDictCursor:
```
Is it possible to give the connections a type hint like you did above?

lukieh2014 replied:
good spot, just added any missing type hints.

lukieh2014 commented:
Added ability for user to unsubscribe from all email services, deleting their data from the RDS. Also added error handling e.g. subscriber email already exists on database, failing connection to RDS. Added logging too, as per Emily's suggestion, and added any missing type hints as Gabriel pointed out.

ebradley12 commented on dashboard/dashboard.py:
```diff
@@ -34,35 +168,73 @@ def report_download_page() -> None:
+                    "No user found with given email address, please try again.")
+            else:
+                st.success(
+                    f"Successfully unsubscribed user with email: {email}")
```
This is a great addition

ebradley12 commented on dashboard/dashboard.py:
```diff
@@ -34,35 +168,73 @@ def report_download_page() -> None:
+            unsubscribed = unsubscribe_user(email)
+            if not unsubscribed:
+                st.error(
+                    "No user found with given email address, please try again.")
```
Good errors

ebradley12 reviewed (approved):
Can't comment much as not much to improve - nice!

emily-curtis reviewed (approved):
great job, with logging docstring etc. Approved!
